src/libstore/s3-binary-cache-store.cc

```diff
@@ -170,6 +174,27 @@ S3Helper::FileTransferResult S3Helper::getObject(
         return Aws::New<std::stringstream>("STRINGSTREAM");
     });
 
+    size_t bytesDone = 0;
+    size_t bytesExpected = 0;
+    request.SetDataReceivedEventHandler([&](const Aws::Http::HttpRequest * req, Aws::Http::HttpResponse * resp, long long l) {
+        if (!bytesExpected && resp->HasHeader("Content-Length")) {
+            if (auto length = string2Int<size_t>(resp->GetHeader("Content-Length"))) {
+                bytesExpected = *length;
+            }
+        }
+        bytesDone += l;
+        act.progress(bytesDone, bytesExpected);
+    });
+
+    request.SetContinueRequestHandler([](const Aws::Http::HttpRequest*) {
+        bool cont = false;
+        try {
+            checkInterrupt();
```

Comment:
This may be problematic since `checkInterrupt()` sets `interruptThrown`, which prevents the interrupt from being thrown on the same thread. Maybe we need a function that checks the interrupt flag without throwing anything, i.e.
```
    return _isInterrupted || (interruptCheck && interruptCheck();
```